### PR TITLE
improve moduleInitializationOrder implementation in dyno, docs

### PR DESCRIPTION
**Pull request by mppf**
The primary focus of this PR is to improve the performance of the `moduleInitializationOrder` query, which is not yet used by the production compiler. It also changes the signature of this function to better match the expected use for it once a later PR wires it up to the production compiler.

The user-facing change is some language spec improvement in the description of module initialization order.

This PR includes a few other minor changes (some of which could impact the production compiler):
 * adjust GatherDecls to avoid delving into submodules
 * improve the language specification description of the module initialization order
 * add a convenience helper `bool idIsModule(Context* context, ID id)`
 * add additional flags to testInteractive to hide the regular output & to skip asking if it should parse again
 * add some tests of the details of module initialization order to `test/modules/initialization`

Reviewed by @DanilaFe - thanks!

- [x] full comm=none testing

### CODE REVIEW

File: frontend/lib/resolution/scope-queries.cpp [hunk 1]
```diff
@@ -2674,131 +2681,137 @@ std::vector<ID> findUsedImportedIds(Context* context,
...
+
+  auto ast = parsing::idToAst(context, modId);
+  if (ast) {
+    if (auto mod = ast->toModule()) {
+      GatherUsedImportedIds gatherModules(context, mod);
+      ast->traverse(gatherModules);
+      // swap the gathered vector in to place
+      ret.swap(gatherModules.idVec);
+    }
   }
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:2756]: If you care about query projection-ness (it kinda sounded like you did in prior PRs?), wouldn't `idToTag(context, modId) == MODULE` here, followed by an (unchecked) `idToAst(...)->toModule()` improve memoization behavior?

File: frontend/lib/resolution/scope-queries.cpp [hunk 2]
```diff
@@ -2674,131 +2681,137 @@ std::vector<ID> findUsedImportedIds(Context* context,
...
+                                   std::vector<ID>& out) {
+  CHPL_ASSERT(!modId.isEmpty() && parsing::idIsModule(context, modId));
 
   // Avoid a cycle walking modules we've seen before.
-  if (seen.find(idMod) != seen.end()) return;
+  if (seen.find(modId) != seen.end()) return;
 
   // We are at least visiting ourselves.
-  seen.insert(idMod);
+  seen.insert(modId);
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:2770]: Isn't the 'insert-and-check' idiom applicable here?

File: frontend/test/resolution/testInteractive.cpp [hunk 3]
```diff
@@ -129,9 +130,11 @@ computeAndPrintStuff(Context* context,
...
   for (const AstNode* child : ast->children()) {
     computeAndPrintStuff(context, child, inFn, calledFns,
-                         scopeResolveOnly, maxIdWidth);
-    if (child->isModule() || child->isFunction()) {
-      std::cout << "\n";
+                         scopeResolveOnly, maxIdWidth, quiet);
+    if (!quiet) {
+      if (child->isModule() || child->isFunction()) {
+        std::cout << "\n";
+      }
```
Comment (DanilaFe) [inline frontend/test/resolution/testInteractive.cpp:137]: Are you doing this for 'performance' measurement or for convenience of use? If it's the latter, consider threading through a "destination stream" (`std::cout` by default), and simply passing in a no-op stream ([my Googling showed me this](https://stackoverflow.com/questions/11826554/standard-no-op-output-stream)) if the output should be silenced. This way, you can avoid all this conditional logic that's broken up across multiple `if quiet` blocks. I don't think C++ is smart enough to ignore `<<` on a a "no-op stream", though, so it wouldn't help -- as much -- if you care about the performance hits of the prints.

File: test/modules/initialization/init-order-grandparent-nouse.chpl [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  writeln("initializing Grandparent");
+  import A;
+
+  module Parent {
+    writeln("initializing Parent");
+    import B;
+
+    module Child {
+      writeln("initializing Child");
+      use C, D;
```
Comment (DanilaFe) [inline test/modules/initialization/init-order-grandparent-nouse.chpl:11]: The name of this test is confusing to me because here's a `use`.

**Review by DanilaFe — APPROVED** [review]

File: frontend/lib/resolution/scope-queries.cpp (hunk 1 above)
Comment (mppf) [reply frontend/lib/resolution/scope-queries.cpp:2756]: I do care about query projection-ness but in this case, it's a misuse of this function if it's not a module. I'll adjust it to use CHPL_ASSERT to make this clearer.

File: frontend/lib/resolution/scope-queries.cpp (hunk 2 above)
Comment (mppf) [reply frontend/lib/resolution/scope-queries.cpp:2770]: Good point, I will change it to do that.

File: frontend/test/resolution/testInteractive.cpp (hunk 3 above)
Comment (mppf) [reply frontend/test/resolution/testInteractive.cpp:137]: It's a little bit of both. I'm disinclined to switch to the no-op stream right now but it's an interesting trick to know about.

File: test/modules/initialization/init-order-grandparent-nouse.chpl (hunk 4 above)
Comment (mppf) [reply test/modules/initialization/init-order-grandparent-nouse.chpl:11]: renamed and added a comment at the top

File: frontend/test/resolution/testInteractive.cpp [hunk 5]
```diff
@@ -236,6 +242,10 @@ int main(int argc, char** argv) {
       trace = true;
     } else if (0 == strcmp(argv[i], "--scope")) {
       scopeResolveOnly = true;
+    } else if (0 == strcmp(argv[i], "--once")) {
```
Comment (riftEmber) [inline frontend/test/resolution/testInteractive.cpp:245]: Noting (late) that these new options aren't documented, but I think it's ok to live with

**Pull request merged**